### ISSUE TITLE
Handle CFEngine-3.5

### ISSUE THREAD

**Issue by lpefferkorn** [lpefferkorn/cfe-profiler#3]
ExpandPromise() prototype is not the same:

3.4.4:

```
void ExpandPromise(enum cfagenttype agent, const char *scopeid, Promise *pp, void *fnptr, const ReportContext *report_context)
```

3.5:

```
void ExpandPromise(EvalContext *ctx, Promise *pp, PromiseActuator *ActOnPromise, void *param)
```

**Comment by basvandervlies** [comment]
I do not know if there  is support for 3.5, I just compiled the master branch and i have no output.

**Comment by lpefferkorn** [comment]
Hello Bas, at the moment 3.5 is not supported, I will add a notice to the README in the meanwhile

**Comment by nickanderson** [comment]
We're ALL waiting. No pressure ;-)

**Comment by lpefferkorn** [comment]
3.5 support will be tested in test-3.5 branch (not online yet), expected in a few days :)

**Comment by nickanderson** [comment]
very exciting

**Comment by lpefferkorn** [comment]
Preliminary version in https://github.com/lpefferkorn/cfe-profiler/tree/test-3.5 

Warning: same multiple reports bug, under investigation

**Comment by basvandervlies** [comment]
Jus
On 16 jul. 2013, at 23:38, Loic Pefferkorn notifications@github.com wrote:

> Preliminary version in https://github.com/lpefferkorn/cfe-profiler/tree/test-3.5
> 
> Warning: same multiple reports bug, under investigation

Just did a test. Still no output. Would thank you for your effort
root# LD_PRELOAD=/var/tmp/bas/cfe-profiler-3.5/cfe_profiler.so cf-agent -KI -DHourly

---

SURFsara has a new telephone number: +31 20 800 1300. 

Bas van der Vlies
| Operations, Support & Development | SURFsara | Science Park 140 | 1098 XG  Amsterdam
| T +31 (0) 20 800 1300  | bas.vandervlies@surfsara.nl | www.surfsara.nl |

**Comment by lpefferkorn** [comment]
@basvandervlies Could you please test against latest test-3.5 branch ?

**Comment by basvandervlies** [comment]
On 18 jul. 2013, at 00:10, Loic Pefferkorn notifications@github.com
 wrote:

> @basvandervlies Could you please test against latest test-3.5 branch ?

of course. Just tested it:
root# LD_PRELOAD=/var/tmp/bas/cfe-profiler-3.5/cfe_profiler.so cf-agent -KI -DHourly
2013-07-18T07:21:53+0200     info: Skipping single-copied file '/etc/ssh/shosts.equiv'
2013-07-18T07:21:53+0200     info: Skipping single-copied file '/etc/ssh/ssh_config'

---

SURFsara has a new telephone number: +31 20 800 1300. 

Bas van der Vlies
| Operations, Support & Development | SURFsara | Science Park 140 | 1098 XG  Amsterdam
| T +31 (0) 20 800 1300  | bas.vandervlies@surfsara.nl | www.surfsara.nl |

**Comment by neilhwatson** [comment]
No seg fault! And output:
**\* Sorted by CPU time - total CPU time: 202.75s ***
    %  time(s) Namespace            Type               Bundle
33.35    67.61   default           agent             efl_main
33.11    67.12   default           agent          efl_bug2638
....

**Comment by lpefferkorn** [comment]
Does that mean it works flawlessly ? :)

**Comment by neilhwatson** [comment]
It doesn't crash and gives a single report. I can't check the math. An explanation on the math would be useful. For example if a bundle calls many bundles is the calling bundle's time an accumulation of all of the bundles called?

**Comment by lpefferkorn** [comment]
There is something definitively wrong. I will have a look.

**Comment by basvandervlies** [comment]
Thanks again.  Can i test something for you?

On 18 jul. 2013, at 20:13, Loic Pefferkorn notifications@github.com wrote:

> There is something definitively wrong. I will have a look.
> 
> —
> Reply to this email directly or view it on GitHub.

---

SURFsara has a new telephone number: +31 20 800 1300. 

Bas van der Vlies
| Operations, Support & Development | SURFsara | Science Park 140 | 1098 XG  Amsterdam
| T +31 (0) 20 800 1300  | bas.vandervlies@surfsara.nl | www.surfsara.nl |

**Comment by lpefferkorn** [comment]
@basvandervlies yes :) Is your cf-agent binary statically linked ? (for the case where you have no statistics at all)

**Comment by basvandervlies** [comment]
sorry for the late response:

root# ldd /usr/sbin/cf-agent 
    linux-vdso.so.1 =>  (0x00007fffd85ec000)
    libdl.so.2 => /lib/libdl.so.2 (0x00002b9f44ce1000)
    libxml2.so.2 => /usr/lib/libxml2.so.2 (0x00002b9f44ee5000)
    libnss_nis.so.2 => /lib/libnss_nis.so.2 (0x00002b9f45237000)
    librt.so.1 => /lib/librt.so.1 (0x00002b9f45442000)
    libm.so.6 => /lib/libm.so.6 (0x00002b9f4564a000)
    libtokyocabinet.so.8 => /usr/lib/libtokyocabinet.so.8 (0x00002b9f458cd000)
    libpcre.so.3 => /lib/libpcre.so.3 (0x00002b9f45b5f000)
    libcrypto.so.0.9.8 => /usr/lib/libcrypto.so.0.9.8 (0x00002b9f45d8f000)
    libacl.so.1 => /lib/libacl.so.1 (0x00002b9f46131000)
    libattr.so.1 => /lib/libattr.so.1 (0x00002b9f46338000)
    libpthread.so.0 => /lib/libpthread.so.0 (0x00002b9f4653c000)
    libc.so.6 => /lib/libc.so.6 (0x00002b9f46759000)
    /lib64/ld-linux-x86-64.so.2 (0x00002b9f44ac1000)
    libz.so.1 => /usr/lib/libz.so.1 (0x00002b9f46abb000)
    libnsl.so.1 => /lib/libnsl.so.1 (0x00002b9f46cd2000)
    libnss_files.so.2 => /lib/libnss_files.so.2 (0x00002b9f46eeb000)
    libbz2.so.1.0 => /lib/libbz2.so.1.0 (0x00002b9f470f8000)

On 18 jul. 2013, at 21:03, Loic Pefferkorn notifications@github.com wrote:

> @basvandervlies yes :) Is your cf-agent binary statically linked ? (for the case where you have no statistics at all)
> 
> —
> Reply to this email directly or view it on GitHub.

---

SURFsara has a new telephone number: +31 20 800 1300. 

Bas van der Vlies
| Operations, Support & Development | SURFsara | Science Park 140 | 1098 XG  Amsterdam
| T +31 (0) 20 800 1300  | bas.vandervlies@surfsara.nl | www.surfsara.nl |

**Comment by lpefferkorn** [comment]
@basvandervlies I don't see any     libpromises.so in your listing, this is why cfe-profiler does not work for you. The LD_PRELOAD trick uses the fact that the function ExpandPromise() is located in libpromises.so to override it.

Is your CFEngine binary a custom build, or a package from a Linux distribution ?

**Comment by lpefferkorn** [comment]
@neilhwatson The time taken by each bundle is measured using wall-clock time. If a bundle calls others bundles, then the time taken by the others bundle will be added to the calling bundle.

l will update the README with an example.

**Comment by basvandervlies** [comment]
Loic,

 Yes i build my own debian packages and the binaries are compiled statically, due to errors in the past when we did cfengine upgrade fails due to differences in libpromises.  I will create a debian package with shared lib, so i can test your software.

regards

On 22 jul. 2013, at 20:07, Loic Pefferkorn notifications@github.com wrote:

> @basvandervlies I don't see any libpromises.so in your listing, this is why cfe-profiler does not work for you. The LD_PRELOAD trick uses the fact that the function ExpandPromise() is located in libpromises.so to override it.
> 
> Is your CFEngine binary a custom build, or a package from a Linux distribution ?
> 
> —
> Reply to this email directly or view it on GitHub.

---

SURFsara has a new telephone number: +31 20 800 1300. 

Bas van der Vlies
| Operations, Support & Development | SURFsara | Science Park 140 | 1098 XG  Amsterdam
| T +31 (0) 20 800 1300  | bas.vandervlies@surfsara.nl | www.surfsara.nl |

**Comment by lpefferkorn** [comment]
Fixed in version 0.2 https://github.com/lpefferkorn/cfe-profiler/releases/tag/v0.2
